source.go

```diff
@@ -8,11 +8,147 @@ import (
+	case src.Context != nil:
+		st, err := src.Context.AsState(&src, sOpt, opts...)
+		if err != nil {
+			return llb.Scratch(), false, nil
+		}
+		return st.With(filter(src.Path, []string{}, []string{})), true, nil
+
+	case src.DockerImage != nil:
+		st, err := src.DockerImage.AsState(name, &src, sOpt, opts...)
+		if err != nil {
+			return llb.Scratch(), false, err
+		}
+
+		var pathHandledByFilter bool
+		if src.DockerImage.Cmd != nil {
+			st = st.With(filter("/", src.Includes, src.Excludes))
+			pathHandledByFilter = false
+		} else {
+			pathHandledByFilter = true
+			st = st.With(filter(src.Path, src.Includes, src.Excludes))
+		}
+		return st, pathHandledByFilter, nil
+
+	case src.Build != nil:
+		st, err := src.Build.AsState(name, &src, filter, sOpt, opts...)
+		if err != nil {
+			return llb.Scratch(), false, err
+		}
+
+		return st.With(filter(src.Path, src.Includes, src.Excludes)), true, nil
+	}
+
+	return llb.Scratch(), false, errNoSourceVariant
+}
+
+func (src *SourceContext) AsState(parent *Source, sOpt SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error) {
+	st, err := sOpt.GetContext(src.Name, localIncludeExcludeMerge(parent.Includes, parent.Excludes), withConstraints(opts))
+	if err != nil {
+		return llb.Scratch(), err
+	}
+
+	if st == nil {
+		return llb.Scratch(), errors.Errorf("context %q not found", src.Name)
+	}
+
+	return *st, nil
+}
+
+func (src *SourceGit) AsState(opts ...llb.ConstraintsOpt) (llb.State, error) {
+	ref, err := gitutil.ParseGitRef(src.URL)
+	if err != nil {
+		return llb.Scratch(), fmt.Errorf("could not parse git ref: %w", err)
+	}
+
+	var gOpts []llb.GitOption
+	if src.KeepGitDir {
+		gOpts = append(gOpts, llb.KeepGitDir())
+	}
+	gOpts = append(gOpts, withConstraints(opts))
+
+	st := llb.Git(ref.Remote, src.Commit, gOpts...)
+	return st, nil
+	// TODO: Pass git secrets
+}
+
+func (src *SourceDockerImage) AsState(name string, parent *Source, sOpt SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error) {
+	bklog.L.Printf("DockerImage: %v", src)
+
+	st := llb.Image(src.Ref, llb.WithMetaResolver(sOpt.Resolver), withConstraints(opts))
+	if src.Cmd == nil {
+		return st, nil
+	}
+
+	if parent.Path == "" {
+		parent.Path = "/"
+	}
+
+	st, err := generateSourceFromImage(name, st, src.Cmd, sOpt, parent.Path, opts...)
+	if err != nil {
+		return llb.Scratch(), err
+	}
+
+	return st, nil
+}
+
+func DefaultSourceFilter(extract string, includes, excludes []string, opts ...llb.ConstraintsOpt) llb.StateOption {
+	return filterState(extract, includes, excludes, opts...)
+}
+
+func TargetMountSourceFilter(_ string, includes, excludes []string, opts ...llb.ConstraintsOpt) llb.StateOption {
+	return filterState("/", includes, excludes, opts...)
+}
+
+func (src *SourceBuild) AsState(name string, _ *Source, filter FilterFunc, sOpt SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error) {
+	st, _, err := GetSource(src.Source, name, filter, sOpt, opts...)
```

Comment:
I'm not sure I'm following why we are passing around a filter function.
The filter should be understood by `GetSource`, and the filter from a parent should not affect the value of `src.Source` but rather the `llb.State` returned by this function.

Reply:
Ah this is a bug. The filter passed in the recursive call should be no-op

Reply:
I'm changing things to not pass the filter through and have the caller apply the filter instead